pkg/gadgets/socket-collector/tracer/bpf/tcp4-collector.c

```diff
@@ -0,0 +1,110 @@
+// SPDX-License-Identifier: GPL-2.0 WITH Linux-syscall-note
+
+/* Copyright (c) 2021 The Inspektor Gadget authors */
+
+/*
+ * Inspired by the BPF selftests in the Linux tree:
+ * https://github.com/torvalds/linux/blob/v5.13/tools/testing/selftests/bpf/progs/bpf_iter_tcp4.c
+ */
+
+/*
+ * This BPF program uses the GPL-restricted function bpf_seq_printf().
+ */
+
+#include <vmlinux/vmlinux.h>
+
+#include <bpf/bpf_helpers.h>
+/*
+ * libbpf v0.4.0 introduced BPF_SEQ_PRINTF in bpf_tracing.h.
+ * In future versions, it will be in bpf_helpers.h.
+ */
+#include <bpf/bpf_tracing.h>
+#include <bpf/bpf_endian.h>
+#include "bpf_tracing_net.h"
+
+char _license[] SEC("license") = "GPL";
+
+static int dump_tcp_sock(struct seq_file *seq, struct tcp_sock *tp)
+{
+	const struct inet_connection_sock *icsk;
+	const struct inet_sock *inet;
+	const struct sock *sp;
+	__u16 destp, srcp;
+	__be32 dest, src;
+	int state;
+
+	icsk = &tp->inet_conn;
+	inet = &icsk->icsk_inet;
+	sp = &inet->sk;
+
+	dest = bpf_ntohl(inet->inet_daddr);
+	src = bpf_ntohl(inet->inet_rcv_saddr);
+	destp = bpf_ntohs(inet->inet_dport);
+	srcp = bpf_ntohs(inet->inet_sport);
+	state = sp->sk_state;
+
+	BPF_SEQ_PRINTF(seq, "tcp %08X %04X %08X %04X %02X\n",
+				src, srcp, dest, destp, state);
+
+	return 0;
+}
+
+static int dump_tw_sock(struct seq_file *seq, struct tcp_timewait_sock *ttw)
+{
+	struct inet_timewait_sock *tw = &ttw->tw_sk;
+	__u16 destp, srcp;
+	__be32 dest, src;
+
+	dest = bpf_ntohl(tw->tw_daddr);
+	src  = bpf_ntohl(tw->tw_rcv_saddr);
+	destp = bpf_ntohs(tw->tw_dport);
+	srcp  = bpf_ntohs(tw->tw_sport);
+
+	BPF_SEQ_PRINTF(seq, "tcp %08X %04X %08X %04X %02X\n",
+				src, srcp, dest, destp, tw->tw_substate);
+
+	return 0;
+}
+
+static int dump_req_sock(struct seq_file *seq, struct tcp_request_sock *treq)
+{
+	struct inet_request_sock *irsk = &treq->req;
+
+	BPF_SEQ_PRINTF(seq, "tcp %08X %04X %08X %04X %02X\n",
+		       bpf_ntohl(irsk->ir_loc_addr),
+		       bpf_ntohs(irsk->ir_num), bpf_ntohl(irsk->ir_rmt_addr),
+		       bpf_ntohs(irsk->ir_rmt_port), TCP_SYN_RECV);
```

Comment:
very minor nit: in other places you're creating temporary variables. Here it's the endianess conversions are done inline with the `BPF_SEQ_PRINTF`. Can you uniform the code?